test/FunctionalTests/Microsoft.OData.Client.Tests/Tracking/ClientEntityTrackerTests.cs

```diff
@@ -0,0 +1,225 @@
+            {
+                await dataServiceContext.SaveChangesAsync();
+            }
+        }
+
+        private void SetupContextWithRequestPipelineForSaving(DataServiceContext[] dataServiceContexts, string response, string location)
+        {
+            foreach (var context in dataServiceContexts)
+            {
+                context.Configurations.RequestPipeline.OnMessageCreating =
+                    (args) => new CustomizedRequestMessage(
+                        args,
+                        response,
+                        new Dictionary<string, string>()
+                        {
+                            {"Content-Type", "application/json;charset=utf-8"},
+                            { "Location", location },
+                        });
+            }
+        }
+
+        class Container : DataServiceContext
+        {
+            public Container(global::System.Uri serviceRoot) :
+                base(serviceRoot, ODataProtocolVersion.V4)
+            {
+                this.Format.LoadServiceModel = () => CsdlReader.Parse(XmlReader.Create(new StringReader(Edmx)));
+                this.Format.UseJson();
+                this.ExampleObjects = base.CreateQuery<ExampleObject>("ExampleObjects");
+                this.ExampleObjectAs = base.CreateQuery<ExampleObjectA>("ExampleObjectAs");
+                this.ExampleObjectBs = base.CreateQuery<ExampleObjectB>("ExampleObjectBs");
+            }
+            public DataServiceQuery<ExampleObject> ExampleObjects { get; private set; }
+            public DataServiceQuery<ExampleObjectA> ExampleObjectAs { get; private set; }
+            public DataServiceQuery<ExampleObjectB> ExampleObjectBs { get; private set; }
+        }
+    }
+    [Key("Id")]
+    public class ExampleObject : BaseEntityType
+    {
+        public int Id { get; set; }
+        public string Name { get; set; }
+        public bool Equals(ExampleObject other)
+        {
+            if (ReferenceEquals(other, null))
+            {
+                return false;
+            }
+            return other.Id == Id;
+        }
+        public override bool Equals(object obj)
+        {
+            var other = obj as ExampleObject;
+
+            if (ReferenceEquals(other, null))
+            {
+                return false;
+            }
+
+            return other.Id == Id;
+        }
+        public override int GetHashCode()
+        {
+            return Id.GetHashCode();
+        }
+    }
+
+    [Key("Id")]
+    public class ExampleObjectA : BaseEntityType, IEqualityComparer<ExampleObject>
+    {
+        public int Id { get; set; }
+        public string Name { get; set; }
+
+        public bool Equals(ExampleObject x, ExampleObject y)
+        {
+            return x.Id == y.Id;
+        }
+
+        public int GetHashCode(ExampleObject obj)
+        {
+            return obj.Id.GetHashCode();
+        }
+    }
+
+    [Key("Id")]
+    public class ExampleObjectB : BaseEntityType

```

Comment:
Could you also add tests cases for POCO classes that do not inherit from `BaseEntityType`?